crates/ra_ide/src/syntax_highlighting/tests.rs

```diff
@@ -131,3 +131,19 @@ fn test_ranges() {
 
     assert_eq!(&highlights[0].highlight.to_string(), "field.declaration");
 }
+
+#[test]
+fn test_flattening() {
```

Comment:
A good test here would be something that touches `highlight_injection` logic:

```
fn fixture(ra_fixture: &str) {}

fn main() {
  fixture(r#"
trait Foo {
    fn foo() {
        println!("2 + 2 = {}", 4);
    }
}
"#)
}
```

This functionality is currently untested simply because naive flattening in highlight_as_html can't deal with it.

Reply:
I wasn't aware of the `highlight_injection` logic, this is only a special case for rust-analyzer development, right? However, this is a great test for the flattening logic, otherwise, Rust does not seem to contain many nestings other than the attributes from the issue.
The current `highlight_as_html` logic iterates over both, the parsed tokens and the highlight ranges. With the change of the function type to be natively flat, it would be better to only iterate over the highlight ranges and consume the text file range by range. Is there an easy way to do that?

Reply:
yup, this is only for rust-anayzer. And yes, you'd need to iteratete the text: 

```
let text = parse.tree().syntax().to_string();
let mut curr = &text[..];
for range in ranges {
    curr = &curr[range.len..]
}
```